### ISSUE TITLE
Basic Javascript - Rewrite and Expansion

### ISSUE THREAD

**Issue by SaintPeter** [FreeCodeCamp/FreeCodeCamp#3881]
Based on the sorts of questions that I am answering on Bonfires, there are some additional topics that need to be covered in greater depth.

As per my conversation with @BerkeleyTrue and @QuincyLarson on October 17th, I have created a "straw dog" version of a proposed updated Basic Javascript Curriculum.  Due to its complexity, I've broken it out into a Google Sheet.

[Link to Google Sheet](https://docs.google.com/spreadsheets/d/1OfnXh3p0XZ7hCE0EFJOQlA8YTst7MH0GVyGJ43URj4E/edit?usp=sharing)

## Overview
There are several goals
1) Increase the number of topics covered.
2) Break the content into "chunks" of similar content, exploring implications
3) Punctuate these chunks with  by "Blank Slate" challenges I am calling "Checkpoints"

Currently I have speced out about 25 new Waypoints/Checkpoints.

I am looking for feedback on the proposed topics.
## New Waypoints
- [x] Waypoint: Storing Values with the Equal Operator
- [x] Waypoint: Initializing Variables with the Equal Operator
- [x] Waypoint: Understanding Uninitialized Variables
- [x] Waypoint: Understanding Case Sensitivity in Variables
- [x] Increment a Number with Javascript
- [x] Decrement a Number with Javascript
- [x] Waypoint: Find a Remainder with Modulus
- [x] Waypoint: Assignment with Plus Equals
- [x] Waypoint: Assignment with Minus Equals
- [x] Waypoint: Assignment with Times Equals
- [x] Waypoint: Assignment with Divided by Equals
- [ ] Checkpoint:  Variables and Assignement
- [x] Waypoint: Quoting Strings with Single Quotes
- [x] Waypoint: Quoting Strings with Double Quotes
- [x] Waypoint: Escape Sequences in String
- [x] Waypoint: Concatanting strings with the Plus Operator
- [x] Waypoint: Concatanting strings with the Plus Equals Operator
- [x] Waypoint: Constructing Strings with Variables
- [x] Waypoint: Appending Variables to Strings
- [x] Waypoint: Understand String Immutability
- [x] Checkpoint:  Strings
- [ ] Checkpoint:  Arrays
- [x] Waypoint: Passing Values to Functions with Arguments
- [x] Waypoint: Global Scope and Functions
- [x] Waypoint: Local Scope and Functions
- [x] Waypoint: Global vs. Local Scope in Functions
- [x] Waypoint: Context for Functions
- [x] Waypoint: Return a Value from a Function with Return
- [x] Waypoint: Assignment with a Returned Value
- [x] Waypoint: Return Early Pattern
- [ ] Waypoint: Optional Arguments for Functions
- [ ] Checkpoint:  Functions
- [x] Waypoint: Accessing Objects Properties with the Dot Operator
- [x] Waypoint: Accessing Objects Properties with Bracket Notation
- [x] Waypoint: Accessing Objects Properties with Variables
- [x] Waypoint: Using Objects for Lookups
- [x] Waypoint: Introducing JavaScript Object Notation (JSON)
- [x] Waypoint:  Accessing Nested Objects in JSON
- [x] Waypoint:  Accessing Nested Arrays in JSON
- [ ] Waypoint:  Arbitrary Nesting in JSON
- [ ] Checkpoint:  Objects
- [x] Waypoint: Comparison Operators - Equality Test
- [x] Waypoint: Comparison Operators - Strict Equality Test
- [x] Waypoint: Comparison Operators - Inequality
- [x] Waypoint: Comparison Operators - Strict Inequality
- [x] Waypoint: Comparison Operators - Greater Than
- [x] Waypoint: Comparison Operators - Greater Than Equal To
- [x] Waypoint: Comparison Operators - Less Than
- [x] Waypoint: Comparison Operators - Less Than Equal To
- [x] Waypoint: Comparison Operators - Logical And
- [x] Waypoint: Comparison Operators - Logical Or
- [x] Waypoint: Introducing Else Statements
- [x] Waypoint: Introducing Else If Statements
- [x] Waypoint: Chaining If/Else Statements
- [x] Waypoint: Selecting from many options with Switch Statements
- [x] Waypoint: Adding a default option in Switch statements
- [x] Waypoint: Mutltiple Identical Options in Switch Statements
- [x] Waypoint: Replacing If/Else chains with Switch
- [x] Waypoint:  Nesting For Loops
- [ ] Checkpoint:  Conditionals and Loops

## Waypoint Rewrites
- [x] Waypoint: Manipulate Arrays With pop
- [ ] Waypoint: Write Reusable JavaScript with Functions
- [ ] Waypoint: Manipulate JavaScript Objects -- "Deleting Object Properties"

**Comment by ltegman** [comment]
Apologies is this thread isn't the right place to put it, but a suggestion by @bradcodd in #3583 was that we might want to add a note about Javascript's case sensitivity in one of the new or existing early lessons for people who might be coming from environments where case sensitivity it not important. I agree that mentioning this and the javascript camelCase naming convention would probably be good additions to the early javascript curriculum.

**Comment by SaintPeter** [comment]
This is a perfect place for it.  I'll add some additional waypoints.  I may just include camelCase as a style note.   Please link JS content related issues to this issue so I can capture them as I build the new content.

**Comment by DidacticDad** [comment]
In the Waypoint: Adding a Default Option in Switch Statements, the example code is missing a semicolon, either as a test or an omission:

switch (num) {
  case value1:
    statement1 (<-- missing semicolon goes here)
    break;
  case value2:
    statement2;
    break;
...
  default:
    defaultStatement;
}

**Comment by ramirobg94** [comment]
On the " Waypoint: Understanding Uninitialized Variables"

Instructions
Initialize the three variables a, b, and c with 5, 10, and "I am a" respectively so that they will not be undefined.

BUT:

b should be defined and have a value of 15

**Comment by dcofjapan** [comment]
Pretty much everything was easy to follow until javascript. it has been an uphill battle since then. I don't want to have to ask for the answer in help chat or look on the wiki for the answer or search in the issues for the answer. For example there are examples such as on this challenge: http://www.freecodecamp.com/challenges/adding-a-default-option-in-switch-statements
 but statement1 told me nothing as a newcomer. Found out by looking at a past issue. mdn didn't really tell me much either since I needed to use the = sign I think. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch
